modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -0,0 +1,549 @@
+      
+      #fill in constants
+      block.h <- Construct.H.multisite(site.ids[ids], var.names, obs.mean[[t]])
+      block.list[[i]]$H <- block.h
+      block.list[[i]]$constant$H <- which(apply(block.h, 2, sum) == 1)
+      block.list[[i]]$constant$N <- length(f.ind)
+      block.list[[i]]$constant$YN <- length(y.ind)
+      block.list[[i]]$constant$q.type <- q.type
+    }
+  }
+  
+  #return values.
+  block.list.all[[t]] <- block.list
+  return(list(block.list.all = block.list.all, H = H, Y = Y, R = R))
+}
+
+##' @title MCMC_Init
+##' @name  MCMC_Init
+##' @author Dongchen Zhang
+##' 
+##' @param block.list  lists of blocks generated by the `build.block.xy` function.
+##' @param X A matrix contains ensemble forecasts.
+##' @details This function helps create initial conditions for the MCMC sampling.
+##' 
+##' @return It returns the `block.list` object with initial conditions filled in.
+MCMC_Init <- function (block.list, X) {
+  var.names <- unique(attributes(X)$dimnames[[2]])
+  #sample mu.f from X.
+  sample.mu.f <- colMeans(X)
+  for (i in seq_along(block.list)) {
+    #number of observations.
+    num.obs <- length(block.list[[i]]$data$y.censored)
+    #loop over each site within each block
+    for (j in seq_along(block.list[[i]]$sites.per.block)) {
+      #initialize mu.f
+      start <- (block.list[[i]]$sites.per.block[j] - 1) * length(var.names) + 1
+      end <- (block.list[[i]]$sites.per.block[j]) * length(var.names)
+      block.list[[i]]$Inits$X.mod <- c(block.list[[i]]$Inits$X.mod, sample.mu.f[start:end])
+      #initialize X
+      block.list[[i]]$Inits$X <- block.list[[i]]$data$y.censored
+      #initialize Xs
+      block.list[[i]]$Inits$Xs <- block.list[[i]]$Inits$X.mod[block.list[[i]]$constant$H]
+    }
+    #initialize q.
+    #if we want the vector q.
+    if (block.list[[i]]$constant$q.type == 1) {
+      for (j in seq_along(block.list[[i]]$data$y.censored)) {
+        block.list[[i]]$Inits$q <- c(block.list[[i]]$Inits$q, stats::rgamma(1, shape = block.list[[i]]$data$aq[j], rate = block.list[[i]]$data$bq[j]))
+      }
+    } else if (block.list[[i]]$constant$q.type == 2) {
+      #if we want the wishart Q.
+      if ("try-error" %in% class(try(block.list[[i]]$Inits$q <- 
+                                     stats::rWishart(1, df = block.list[[i]]$data$bq, Sigma = block.list[[i]]$data$aq)[,,1], silent = T))) {
+        block.list[[i]]$Inits$q <- 
+          stats::rWishart(1, df = block.list[[i]]$data$bq, Sigma = stats::toeplitz((block.list[[i]]$constant$YN:1)/block.list[[i]]$constant$YN))[,,1]
+      }
+    }
+  }
+  #return values.
+  return(block.list)
+}
+
+##' @title MCMC_block_function
+##' @name  MCMC_block_function
+##' @author Dongchen Zhang
+##' 
+##' @param block  each block within the `block.list` lists.
+##' 
+##' @return It returns the `block` object with analysis results filled in.
+MCMC_block_function <- function(block) {
+  #build nimble model
+  model_pred <- nimble::nimbleModel(GEF.Block.Nimble,
```

Comment:
I remember in Raiho's code there was a way of separating the compilation of the C code from the execution of the nimble model. You might want to look into something like that here so that you don't re-parse and re-compile the nimble code separately for every site or site-block.

Reply:
The previous code could save computation complexity by propagating values into compiled MCMC objects. But that doesn't account for any dynamic changes to the observations, which will require the recompile process of the nimble MCMC. Indeed, we need to specify if we need to recompile the nimble if the observations change over time. But I imagine it to be a separate PR in the future.